File: skylib-c++/src-lib/property_manager.cpp


#include <iostream>
#include <memory>

#include "property_manager.hh"
#include "property.hh"
#include "utils.hh"


namespace ch_skymarshall::gui {

using namespace std;

property_manager::property_manager() = default;

property_manager::~property_manager() DESTR_WITH_LOG("~property_manager")

void property_manager::add_listener(const string_view& _name, shared_ptr<property_listener> _listener) {
	auto iter = m_propertyListeners.find(_name);
	shared_ptr<listener_list_type> plist;
	if (iter == m_propertyListeners.end()) {
		plist = make_shared<listener_list_type>();
		m_propertyListeners[_name] = plist;
	} else {
		plist = (*iter).second;
	}
	plist->push_back(_listener);
}

void property_manager::remove_listener(const string_view& _name, weak_ptr<property_listener> _listener) {
	auto iter = m_propertyListeners.find(_name);
	if (iter == m_propertyListeners.end()) {
		return;
	}
	auto plist = (*iter).second;
	plist->remove(_listener.lock());
	if (plist->empty()) {
		m_propertyListeners.erase(_name);
	}
}

void property_manager::remove_listeners(const string_view &_name) {
	m_propertyListeners.erase(_name);
}

void property_manager::fire_property_changed(source_ptr _source, const string_view &_name, const void* _oldValue,
		const void* _newValue) const {
	auto iter = m_propertyListeners.find(_name);
	if (iter == m_propertyListeners.end()) {
		return;
	}
	auto plist((*iter).second);
	for (auto listener = plist->begin(); listener != plist->end(); listener++) {
		(*listener)->fire(_source, _name, _oldValue, _newValue);
	}
}

void property_manager::fire_before_property_changed(source_ptr _source, property* _property) const {
	auto iter = m_propertyListeners.find(_property->name());
	if (iter == m_propertyListeners.end()) {
		return;
	}
	auto plist((*iter).second);
	for (auto listener = plist->begin(); listener != plist->end(); listener++) {
		(*listener)->before_change(_source, _property);
	}
}

void property_manager::fire_after_property_changed(source_ptr _source, property* _property) const {
	auto iter = m_propertyListeners.find(_property->name());
	if (iter == m_propertyListeners.end()) {
		return;
	}
	auto plist((*iter).second);
	for (auto listener = plist->begin(); listener != plist->end(); listener++) {
		(*listener)->after_change(_source, _property);
	}
}

void property_manager::dump() const {
	for (auto& mapListener: m_propertyListeners) {
		auto plist = mapListener.second;
		for (auto listener = plist->begin(); listener != plist->end(); listener++) {
			cout << "  " << hex << *listener << endl;
		}
	}

}

}

File: skylib-rust/src/lib_properties/property_manager.rs
//! property_manager.rs
//!
//! Rust port of `property_manager.hh` (originally `PropertyManager.hh`).
//!
//! Original author: scaille (Feb 18, 2012)
//!
//! ## Design notes
//! - C++ keyed listeners by `string_view` (a non-owning view into a
//!   longer-lived string, typically a string literal naming the
//!   property). Rust's borrow checker makes that awkward for a
//!   `HashMap` key unless the referenced data truly outlives the map, so
//!   this port keys by `&'static str` instead — a direct fit given
//!   property names in this codebase are always compile-time literals.
//! - `remove_listener(name, weak_ptr)` / `remove_listener(name, void*)`
//!   both existed in C++ to support two ways of identifying a listener
//!   for removal (a still-possibly-alive weak handle, or a raw identity
//!   pointer). Both are kept here as `remove_listener` /
//!   `remove_listener_by_ref`.

use std::rc::{Rc};

use crate::lib_properties::{Property, PropertyManager, PropertyListener, SourcePtr};

impl PropertyManager {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_listener(&self, name: &'static str, listener: Rc<dyn PropertyListener>) {
        self.listeners
            .borrow_mut()
            .entry(name)
            .or_default()
            .push(listener);
    }

    pub fn remove_listeners(&self, name: &'static str) {
        self.listeners.borrow_mut().remove(name);
    }

    /// Removes a listener identified by a (possibly already-expired) weak
    /// handle, comparing by pointer identity once upgraded.
    pub fn remove_listener(&self, name: &'static str, listener: Rc<dyn PropertyListener>) {
        if let Some(list) = self.listeners.borrow_mut().get_mut(name) {
            list.retain(|l| Rc::ptr_eq(l, &listener));
        }
    }

    pub fn fire_property_changed(
        &self,
        source: SourcePtr,
        name: &str,
        old_value: *const (),
        new_value: *const (),
    ) {
        if let Some(list) = self.listeners.borrow().get(name) {
            for listener in list {
                listener.fire(source, name, old_value, new_value);
            }
        }
    }

    pub fn fire_before_property_changed(&self, source: SourcePtr, property: &dyn Property) {
        if let Some(list) = self.listeners.borrow().get(property.name()) {
            for listener in list {
                listener.before_change(source, property);
            }
        }
    }

    pub fn fire_after_property_changed(&self, source: SourcePtr, property: &dyn Property) {
        if let Some(list) = self.listeners.borrow().get(property.name()) {
            for listener in list {
                listener.after_change(source, property);
            }
        }
    }

    pub fn dump(&self) {
        for (name, list) in self.listeners.borrow().iter() {
            println!("{name}: {} listener(s)", list.len());
        }
    }
}
